nat-lab/tests/test_batching.py

```diff
@@ -240,3 +242,77 @@ async def start_node_manually(client, node, sleep_s):
 
             print("Delay chart below")
             print(delay_chart)
+
+
+def proxying_peer_parameters(clients: List[ConnectionTag]):
+    def features():
+        features = default_features(enable_direct=False, enable_nurse=False)
+        features.wireguard.persistent_keepalive.proxying = 5
+        features.link_detection = FeatureLinkDetection(
+            rtt_seconds=2, no_of_pings=0, use_for_downgrade=False
+        )
+
+        features.batching = FeatureBatching(
+            direct_connection_threshold=5,
+            trigger_cooldown_duration=60,
+            trigger_effective_duration=10,
+        )
+        return features
+
+    return [
+        SetupParameters(
+            connection_tag=conn_tag,
+            adapter_type_override=TelioAdapterType.NEP_TUN,
+            features=features(),
+            fingerprint=f"{conn_tag}",
+        )
+        for conn_tag in clients
+    ]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "setup_params",
+    [
+        proxying_peer_parameters(
+            [ConnectionTag.DOCKER_CONE_CLIENT_1, ConnectionTag.DOCKER_CONE_CLIENT_2]
+        )
+    ],
+)
+@pytest.mark.timeout(60)
+async def test_proxying_peers_reconnect(setup_params: List[SetupParameters]) -> None:
+    # Since batching keepalives are performed on application level instead of Wireguard
+    # backend we need to ensure that proxying peers are receiving the keepalives. To test
+    # for that we can enable link detection that guarantees quick detection if there's no corresponding
+    # received traffic(WireGuard PassiveKeepalive). If batcher correctly emits pings, it
+    # should trigger link detection quite quickly.
+    async with AsyncExitStack() as exit_stack:
+        env = await setup_mesh_nodes(exit_stack, setup_params)
+
+        await asyncio.gather(*[
+            await exit_stack.enter_async_context(
+                run_async_context(
+                    client.wait_for_state_peer(
+                        node.public_key, [NodeState.CONNECTED], [PathType.RELAY]
+                    )
+                )
+            )
+            for client, node in itertools.product(env.clients, env.nodes)
+            if not client.is_node(node)
+        ])
+
+        alpha, beta = env.clients
+        await beta.stop_device()
+
+        _, beta_node = env.nodes
+
+        # 20 seconds should be enough since:
+        # * 10 seconds is for WireGuard's Passive-Keepalive
+        # * RTT is configured to be 2 seconds
+        # * no pings to validate
+        # * PersistentKeepalive configured to 5seconds
+        # Given all these it should be worst case latency until detection: 10+2+5=17
+        await asyncio.sleep(20)
```

Comment:
Maybe instead of sleep + assert it would be better to actively wait for LinkState.DOWN? I'm afraid the current approach can cause some unnecessary flakiness.